xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/internal/doc/XWikiAttachmentList.java

```diff
@@ -0,0 +1,250 @@
+        for (XWikiAttachment x : c) {
+            map.put(x.getFilename(), x);
+            added(x);
+        }
+        updateMap();
+        return true;
+    }
+
+    /**
+     * Adds all attachments to the list in order of filename.
+     * 
+     * @param index index is ignored as list is reordered based on filename
+     * @param c Collection that contains XWikiAttachment objects
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean addAll(int index, Collection<? extends XWikiAttachment> c)
+    {
+        return addAll(c);
+    }
+
+    /**
+     * {@inheritDoc}
+     * 
+     * @since 10.0RC1
+     */
+    @Override
+    public XWikiAttachment remove(int index)
+    {
+        XWikiAttachment removedAttachment = map.remove(this.list.get(index).getFilename());
+        if (removedAttachment != null) {
+            removedAttachment = super.remove(index);
+            document.setMetaDataDirty(true);
+        }
+        return removedAttachment;
+    }
+
+    /**
+     * Removes XWikiAttachment.
+     * 
+     * @param attachment XWikiAttachment to remove.
+     * @return true unless the attachment is not found
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean remove(Object attachment)
+    {
+        XWikiAttachment xwikiAttachment = (XWikiAttachment) attachment;
+        if(map.put(xwikiAttachment.getFilename(), xwikiAttachment) == null)
+            return false;
+        map.remove(attachment);
+        updateMap();
+        return true;
+    }
+
+    /**
+     * Adds or replaces attachment with the same filename as the parameter.
+     * 
+     * @param attachment the attachment to add to the list
+     * @return the attachment that was added to the list in order of filename
+     * @since 10.0RC1
+     */
+    public XWikiAttachment set(XWikiAttachment attachment)
+    {
+        map.put(attachment.getFilename(), attachment);
+        added(attachment);
+        updateMap();
+        return attachment;
+    }
+
+    /**
+     * Adds or replaces attachment with the same filename as the parameter.
+     * 
+     * @param index this parameter is not used but is needed to override the method
+     * @param attachment the attachment to add to the list
+     * @return the attachment that was added to the list in order of filename
+     * @since 10.0RC1
+     */
+    @Override
+    public XWikiAttachment set(int index, XWikiAttachment attachment)
+    {
+        return set(attachment);
+    }
+
+    /**
+     * @param filename the filename of the attachment to be returned.
+     * @return attachment with the given filename or null if not found.
+     * @since 10.0RC1
+     */
+    public XWikiAttachment getByFilename(String filename)
+    {
+        return map.get(filename);
+    }
+
+    @Override
+    public boolean removeAll(Collection<?> c)
+    {
+        boolean changed = false;
+        for (XWikiAttachment x : (Collection<? extends XWikiAttachment>) c) {
+            if (this.list.contains(x)) {
```

Comment:
You don't need this like since `remove()` gives you the same information already.